src/dd/dd.rs

```diff
@@ -0,0 +1,640 @@
+                {
+                    if m < cf.out_block_size {
+                        cf.out_part += 1;
+                    } else {
+                        cf.out_full += 1;
+                    }
+                }
+                pos += m;
+            } else {
+                err("error occured".to_string()); // TODO: handle
+            }
+        }
+        if total >= cf.inp_block_size * cf.count {
+            break;
+        }
+    }
+    display_stat(&cf);
+}
+
+enum InpFileType {
+    Reg(File),
+    Stdin(io::Stdin),
+}
+
+pub struct InpFileWrap {
+    f: InpFileType,
+    is_seekable: bool,
+}
+
+impl InpFileWrap {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        match self.f {
+            InpFileType::Reg(ref mut f) => f.read(buf),
+            InpFileType::Stdin(ref mut s) => s.read(buf),
+        }
+    }
+}
+
+enum OutFileType {
+    Reg(File),
+    Stdout(io::Stdout),
+}
+
+pub struct OutFileWrap {
+    f: OutFileType,
+    is_seekable: bool,
+}
+
+impl OutFileWrap {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        match self.f {
+            OutFileType::Reg(ref mut f) => f.write(buf),
+            OutFileType::Stdout(ref mut s) => s.write(buf),
+        }
+    }
+}
+
+pub fn uumain(args: Vec<String>) -> i32 {
+    let mut cf = parse_args();
+    //    println!("{}", cf);
+    setup(&mut cf);
+    0
+    //    let signal = chan_signal::notify(&[Signal::INT]);
+    //    let (s_done, r_done) = chan::sync(0);
+    //
+    //    let cf = parse_args();
+    //    let cf2 = Arc::clone(&cf);
+    //
+    //    let handle = thread::spawn(move || run(cf2, s_done));
+    //
+    //    chan_select! { // FIXME: resume copying after printing progress
+    //        signal.recv() -> signal => display_progress(Arc::clone(&cf)),
+    //        r_done.recv() => ()
+    //    }
+    //    handle.join();
+}
+
+pub fn get_u(val: &str) -> Option<usize> {
+    if val.is_empty() || (val.len() == 1 && !val.chars().nth(0).unwrap().is_digit(10))
+        || val.chars().nth(0).unwrap() == '-'
+        {
+            return None;
+        }
+
+    let mut pref = val;
+    let mut size: usize = match val.chars().nth(val.len() - 1) {
+        Some('k') => {
+            pref = &val[..val.len() - 1];
+            1024
+        }
+        Some('b') => {
+            pref = &val[..val.len() - 1];
+            512
+        }
+        Some(m) if !m.is_digit(10) && m != 'k' && m != 'b' => {
+            return None;
+        }
+        _ => 1,
+    };
+
+    if pref.is_empty() || pref.chars().nth(pref.len() - 1).unwrap() == 'x'
+        || pref.chars().nth(0).unwrap() == 'x'
+        {
+            return None;
+        }
+    for m in pref.split('x').collect::<Vec<&str>>() {
+        if let Ok(n) = m.parse::<usize>() {
+            size *= n;
+        } else {
+            return None;
+        }
+    }
+
+    Some(size)
+}
+
+pub fn get_i(val: &str) -> Option<isize> {
```

Comment:
This function is very similar to `get_u()`, so you should probably factor out the common code into a function or trait.  Also, I'm not 100% sure, but there may be crates on [crates.io](https://crates.io) that can do this for you.

Reply:
Yeah as far as I can tell there is only one line that is meaningfully different (the first `if` expression).